### ISSUE TITLE
Apparently CJDNS network does not work with Tor on mainnet.

### ISSUE THREAD

**Issue by lsilva01** [bitcoin/bitcoin#24450]
Apparently CJDNS network does not work with Tor on mainnet.

**Expected behavior**

Both networks should work.

**Actual behavior**

When monitoring connections via `./src/bitcoin-cli  -netinfo 4`, this does not show any CJDNS peers if Tor is enabled.
If the Tor proxy is commented out in the configuration file, the CJDNS peers will be connected.

**To reproduce**

Install Tor, I2P and CJDNS services on Fedora / Debian.

Add the following items to `bitcoin.conf` file:
```
[main]
proxy=127.0.0.1:9050
cjdnsreachable=1
i2pacceptincoming=1
i2psam=127.0.0.1:7656
addnode=fcc7:be49:ccd1:dc91:3125:f0da:457d:8ce
addnode=fc32:17ea:e415:c3bf:9808:149d:b5a2:c9aa
addnode=sedndhv5vpcgdmykyi5st4yqhdxl3hpdtglta4do435wupahhx6q.b32.i2p:0
addnode=zsxwyo6qcn3chqzwxnseusqgsnuw3maqnztkiypyfxtya4snkoka.b32.i2p:0
addnode=wwbw7nqr3ahkqv62cuqfwgtneekvvpnuc4i4f6yo7tpoqjswvcwa.b32.i2p:0
addnode=opnyfyeiibe5qo5a3wbxzbb4xdiagc32bbce46owmertdknta5mi7uyd.onion:8333
addnode=rp7k2go3s5lyj3fnj6zn62ktarlrsft2ohlsxkyd7v3e3idqyptvread.onion:8333
addnode=xqt25cobm5zqucac3634zfght72he6u3eagfyej5ellbhcdgos7t2had.onion:8333
```
Run Bicoin Core. With the configuration above, there can be IPv4, onion and I2P peers, but not CJDNS.
Stop Bitcoin Core, comment out `proxy=127.0.0.1:9050` and restart it. There can be IPv4, CJDNS and I2P

**System information**
Bitcoin Core was compiled from master (last commit `ee8c9971256`). Ran on Fedora Linux 35. 
Same behavior observed on Debian 11.

**Comment by laanwj** [comment]
I think this is at least partially expected behavior. The thing is, if you're proxying IPv6 over Tor, CJDNS will not work. `-proxy` is the catch-all.

Generally if you want to run a multi-network node, set a proxy per network, e.g. `-onion=<ip:port>`. This will cause non-CJDNS IPv4 and IPv6 to go over clearnet, though, so might not be the configuration you want.

**Comment by jonatack** [comment]
> Apparently CJDNS network does not work with Tor on mainnet.
> 
> this does not show any CJDNS peers if Tor is enabled

A few ways I have run CJDNS with Tor (provided it does what a user wants):

- with a proxy per network as suggested by @laanwj 
- without setting a proxy
- with `-proxy` and `-discover` (and optionally `-listen`)
- with `-onion=127.0.0.1:9050 -externalip=<your onion address> -discover -listen`

**Comment by vasild** [comment]
Yeah, I would say the above is misconfiguration which can be fixed by replacing `-proxy=` with `-onion=`. Some thoughts:

`-proxy=` is catch-all setting. There may be a proxy running on one's router that also handles CJDNS. So, it would be undesirable to change the behavior to ignore `-proxy=` within Bitcoin Core when connecting to CJDNS addresses.

`-proxy=` is ignored for I2P addresses, so it is catch-all _except_ I2P.

Looks like we don't support right now a setup where IPv4 and IPv6 are routed via Tor and CJDNS connections are made without a proxy.

Maybe we should consider something like `-proxy=1.2.3.4:5678=ipv4` for a more fine-grained control?

**Comment by felipelalli** [comment]
I think this should be better documented in https://github.com/bitcoin/bitcoin/blob/master/doc/cjdns.md .

**Comment by pinheadmz** [comment]
I could work on this. I think it would be cool to support all networks with full privacy (meaning ipv4 and ipv6 routed over Tor). Onion goes over the SOCKS5 and I2P has its own proxy server (SAM), CJDNS is just a little trickier.

> Maybe we should consider something like `-proxy=1.2.3.4:5678=ipv4` for a more fine-grained control?

I think it's either something like this, or another option like `-cjdnsproxy` (default true if `-proxy` is set)

**Comment by pinheadmz** [comment]
No PR yet but I am looking into which solution is best. Any thoughts ?

**Comment by vasild** [comment]
Taking a step back from `-proxy=1.2.3.4:5678=ipv4` vs `-cjdnsproxy=`: if very few people use CJDNS for Bitcoin, is it worth spending effort on CJDNS at all? On my node there are 3 CJDNS addresses in addrman:

```sh
$ bitcoin-cli getpeerinfo |jq 'map(select(.network == "cjdns")) |length'
3
```

**Comment by caesrcd** [comment]
I’ve been studying the multi-network support in Bitcoin Core (Clearnet, Tor, I2P, CJDNS), especially from a resilience and redundancy perspective. I’d like to raise a concern regarding the current behavior of routing CJDNS through the general `-proxy` setting.

While it's understandable that `-proxy` applies broadly to IPv4, IPv6, and CJDNS, this design unintentionally breaks a core benefit of multi-network support: **redundancy through independence**.

Unlike Tor or I2P, which are anonymity-focused overlays, CJDNS operates as a self-contained IPv6 mesh network (`fc00::/8`). It doesn't work over SOCKS5, and attempting to route it via Tor’s proxy fails silently or stalls connections — which seems to be what users are reporting.

More importantly, even if it did work, proxying CJDNS over Tor `eliminates its redundancy value`. If CJDNS depends on Tor, it no longer provides an independent path for Bitcoin traffic when Tor is down, censored, or degraded.

**I suggest:**

- Excluding CJDNS from the general `-proxy` path by default (like is already done for I2P).
- Optionally allowing `-cjdnsproxy=` for advanced setups, but defaulting to direct connection.

This would better align with Bitcoin Core’s goal of maximizing connectivity robustness, especially in adversarial or degraded network conditions.
